### PR TITLE
Huffman Coding

### PR DESCRIPTION
**Pull request by lalith1403**
I have added an article on Huffman Coding.

short_description: An article that introduces the concept of huffman coding and a step by step tutorial to implement Python code to generate Huffman codes for a given string

### CODE REVIEW

**Comment by hectorkambow** [comment]
Plagiarism check done

**Comment by nadivgold** [comment]
@hectorkambow can you take a look over and correct the description and first paragraph?

File: articles/huffman-coding-in-python/index.md [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+    alt: huffman python code coding data text compression
+---
+Huffman coding is a lossless way to compress and encode text based on the frequency of the characters in the text.
+
+<!--more-->
+# Huffman Coding
+
+In computer science and information theory, [Huffman code](https://www.sciencedirect.com/topics/computer-science/codeword-length) is a special type of optimal prefix code that is often used for lossless data compression. The algorithm was developed by [David A. Huffman](https://en.wikipedia.org/wiki/Huffman_coding) in the late 19th century as part of his research into computer programming and is commonly found in programming languages such as C, C + +, [Java](https://algs4.cs.princeton.edu/55compression/Huffman.java.html), JavaScript, Python, Ruby, etc. The intuition behind Huffman encoding is as follows. A letter or a symbol that occurs frequently is represented by a shorter code and a symbol that occurs rarely is represented by a longer code. This leads to efficient representation of characters that require less memory to be stored in. Therefore, we can find Huffman coding used in data compression techniques.
```
Comment (nadivgold) [inline articles/huffman-coding-in-python/index.md:22]: Ruby, etc. The intuition ->  Ruby, and more. The intuition

Comment (nadivgold) [reply articles/huffman-coding-in-python/index.md:22]: as follows. A letter -> as follows: a letter

Comment (nadivgold) [reply articles/huffman-coding-in-python/index.md:22]: shorter code and a symbol -> shorter code, and a symbol

Comment (nadivgold) [reply articles/huffman-coding-in-python/index.md:22]: and a symbol that occurs -> and a letter or symbol that occurs

Comment (nadivgold) [reply articles/huffman-coding-in-python/index.md:22]: stored in. Therefore, we can find -> stored in. This means 

 Huffman coding used in data ->  Huffman coding can be used as a data

 data compression techniques. ->  data compression technique.

File: articles/huffman-coding-in-python/index.md [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+
+In this article, we are going to cover the following:
+   
+1. Theory of Huffman coding
+2. Code to compute Huffman code for a given string
```
Comment (nadivgold) [inline articles/huffman-coding-in-python/index.md:27]: compute Huffman code for -> compute the Huffman code for

File: articles/huffman-coding-in-python/index.md [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+   
+### Theory of Huffman Coding
+
+The algorithm is based on the frequency with which each character in the file appears and the number of characters in a data structure with a frequency of 0. The Huffman encoding for a typical text file saves about 40%. We know that a file is stored on a computer as binary code and that each character in the file has been assigned a binary character, and the character codes usually have a fixed length for different characters. Huffman binary code, such as compiled executables, would therefore have a different space-saving. A binary file in which ASCII character is encoded with a frequency of 0.5 would have a very different distribution and frequency from its ASCII counterpart.
```
Comment (nadivgold) [inline articles/huffman-coding-in-python/index.md:31]: The algorithm is based -> Huffman Coding is based

Comment (nadivgold) [reply articles/huffman-coding-in-python/index.md:31]: saves about 40%. We know that ->  saves about 40% of the size of the original data. We know that

Comment (nadivgold) [reply articles/huffman-coding-in-python/index.md:31]: binary code and that each character -> binary code, and that each character

Comment (nadivgold) [reply articles/huffman-coding-in-python/index.md:31]: and the character codes ->  and character codes

Comment (nadivgold) [reply articles/huffman-coding-in-python/index.md:31]: which ASCII character -> which an ASCII character

File: articles/huffman-coding-in-python/index.md [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+
+To compress a file with a sequence of characters, we need a table that gives us the sequences of bits used for each character. This table creates an encoding tree that uses the root/leaf path to create a bit sequence that encodes the characters. We can follow the roots and leaves to create a list of all characters with the maximum bit length of the encoded characters and the number of occurrences. 
+
+To construct an optimal tree, we use the greedy algorithm. Huffman encoding tree provides a minimum length of the character encodings for data compression. The nodes in the tree represent the frequency of a character's occurrence. The root node represents the length of the string, and traversing the tree gives us the character-specific encodings. Once the tree is constructed, traversing the tree gives us the respective codes for each symbol.
```
Comment (nadivgold) [inline articles/huffman-coding-in-python/index.md:35]: use the greedy algorithm. -> use a greedy algorithm.

Comment (nadivgold) [reply articles/huffman-coding-in-python/index.md:35]: encoding tree provides a -> encoding trees provides a

File: articles/huffman-coding-in-python/index.md [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+
+The optimal tree upon completion is given in the image below:
+![optimal Tree](optimaltree.jpg)
+
+We will develop and implement a program that implements Huffman coding in the next section.  
+### Code
+
+We first define a class called `HuffmanCode` which is initialized with probabilities. Let us look at the flow of the code implemented below:
+
+1. Obtain the string and compute the frequency
```
Comment (nadivgold) [inline articles/huffman-coding-in-python/index.md:45]: compute the frequency -> compute the frequency of each character in the string

File: articles/huffman-coding-in-python/index.md [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+2. Using the frequency, obtain the probabilities
+3. Using the algorithm, compute the Huffman codes
+4. For the Huffman codes, compute mean length, variance, and entropy
+   
+Let us look at the algorithm used to compute the Huffman codes.
+
+1. We initially sort the probabilities in decreasing order. This is necessary to assign codes to each symbol according to their frequencies.
+2. Run a for loop from `0` to `length of string-2`
```
Comment (nadivgold) [inline articles/huffman-coding-in-python/index.md:53]: Run a for loop -> Run a `for` loop

File: articles/huffman-coding-in-python/index.md [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+3. For the first symbol assign `1` as the default code
+4. For the upcoming symbols, check for the previous symbols
+   1. If 1- add 0
+   2. If 11- add 1
+   3. If 10 - add 0
+   4. If 111 - add 1
+5. Once the Huffman codes are generated, read them in the reverse order to obtain the final_code. This is done to generate code as done by tree traversal.
```
Comment (nadivgold) [inline articles/huffman-coding-in-python/index.md:60]: the final_code. This is done ->  the `final_code`. This is done

File: articles/huffman-coding-in-python/index.md [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+6. final_code is the list that contains all the huffman_codes in the order of the probabilities
```
Comment (nadivgold) [inline articles/huffman-coding-in-python/index.md:61]: final_code is the list ->  `final_code` is the list

File: articles/huffman-coding-in-python/index.md [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+
+Let us look at the code now. There are a couple of functions defined inside the class which will be explained in the section given below.
```
Comment (nadivgold) [inline articles/huffman-coding-in-python/index.md:63]: Let us look at the code now.  -> Let us now look at the code.

Makes the sentence sound less aggressive and threatening.

File: articles/huffman-coding-in-python/index.md [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+                return j
+        return index-1
+
+    def characteristics_huffman_code(self, code):
+        length_of_code = [len(k) for k in code]
+
+        mean_length = sum([a*b for a, b in zip(length_of_code, self.probability)])
+
+        print("Average length of the code: %f" % mean_length)
+        print("Efficiency of the code: %f" % (entropy_of_code/mean_length))
```
Comment (nadivgold) [inline articles/huffman-coding-in-python/index.md:87]: `entropy_of_code` is undefined.

File: articles/huffman-coding-in-python/index.md [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+    if huffman_code[id]=='':
+        print(' %-4r |%12s' % (char[0], 1))
+        continue
+    print(' %-4r |%12s' % (char[0], huffman_code[id]))
+
+huffmanClassObject.characteristics_huffman_code(huffman_code)
+```
+
+#### Auxiliary functions
+We have a couple of auxiliary functions such as find_position and characteristics_huffman_code. find_position is used to insert bits to the existing code computed in the n-3 previous iterations, where n is the length.
```
Comment (nadivgold) [inline articles/huffman-coding-in-python/index.md:173]: such as find_position and characteristics_huffman_code. find_position is  -> such as find_position and `characteristics_huffman_code`. `find_position` is

File: articles/huffman-coding-in-python/index.md [hunk 12]
```diff
@@ -0,0 +1,184 @@
...
+
+The second auxiliary function defined is characteristics_huffman_code. This function generates the mean length of the codes, entropy, variance, and efficiency. Let us look at each of these in detail
```
Comment (nadivgold) [inline articles/huffman-coding-in-python/index.md:175]: characteristics_huffman_code. This function -> `characteristics_huffman_code`. This function

File: articles/huffman-coding-in-python/index.md [hunk 13]
```diff
@@ -0,0 +1,184 @@
...
+
+![huffman code output](Code_Output.jpg)
+### Performance of Huffman Code
+1. Average Length: It is defined as the [average number of bits](https://pdfs.semanticscholar.org/8ee2/a74ee0371c40dd751392762572409767d5db.pdf) required to represent a character in the string. Let us consider an example.
```
Comment (nadivgold) [inline articles/huffman-coding-in-python/index.md:179]: > Let us consider an example.

Can't say this without giving an example.

Comment (nadivgold) [reply articles/huffman-coding-in-python/index.md:179]: Maybe you meant the above example, in which case, it's not explicit enough for the reader to understand.

File: articles/huffman-coding-in-python/index.md [hunk 14]
```diff
@@ -0,0 +1,184 @@
...
+2. Efficiency:  Efficiency is a measure of the number of bits wasted. An efficiency of 0.945 means for every 100 bits 5.5 bits are wasted. This deals with the concept of entropy. Entropy is a quantitative measure for the amount of information a code provides. Any repetition results in redundancy thereby reducing the information per unit symbol. Thus efficiency decreases.
+
+### Conclusion
+
+In this article, we have discussed what Huffman coding is, some of its applications, and have also considered a code that encodes a given text based on the frequency of characters. We hope you have enjoyed the article. We suggest you implement the code and use this guide as a code-along article.
```
Comment (nadivgold) [inline articles/huffman-coding-in-python/index.md:184]: characters. We hope ->  characters. I hope

Comment (nadivgold) [reply articles/huffman-coding-in-python/index.md:184]: article. We suggest -> article. I suggest

**Comment by hectorkambow** [comment]
@lalith1403  - wante to double check that any of the images in the article do not need image sources ???

**Comment by lalith1403** [comment]
> @lalith1403 - wanted to double check that any of the images in the article do not need image sources ???

I have added the required link. One of the images is a screenshot of my screen.

**Comment by ninjaginja** [comment]
@lalith1403 we received an email regarding a potential error in this published article. Could you please investigate and make any needed corrections?

> I read your article at https://www.section.io/engineering-education/huffman-coding-python/. However, the output of the code seems not correct as shown in the following picture.

>截屏2021-01-12 下午3 45 47

>The result is not an instantaneous code because 110 includes 10, and I think it should be the problem of step 4.

**Comment by lalith1403** [comment]
Noted. I will go through the error once. It's weird since it worked fine for my example. Kindly give me a couple of days to rectify the error.

**Comment by ninjaginja** [comment]
sure thing, no problem

**Pull request merged**